### PR TITLE
52224 - create update exchange set

### PR DESCRIPTION
**Pull request by Vishal14583**
- Update exchange set created
- Unit Tests
- Mock API changes
- Functional Tests

### CODE REVIEW

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/TestConfiguration.cs [hunk 1]
```diff
@@ -55,8 +55,11 @@ public class POSFileDetails
             public string M02Sha1File { get; set; }
             public string ZipFilesBatchId { get; set; }
             public string IsoSha1BatchId { get; set; }
+            public string UpdateSetBatchId { get; set; }
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/TestConfiguration.cs]: Done,

**Review by harshal11869 — APPROVED** [review]
LGTM

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 2]
```diff
@@ -7,39 +7,49 @@ namespace UKHO.PeriodicOutputService.API.FunctionalTests.Helpers
     public static class FssBatchHelper
     {
         private static FssApiClient FssApiClient { get; set; }
+        static FSSApiConfiguration FSSAuth = new TestConfiguration().FssConfig;
+        private static readonly POSFileDetails posDetails = new TestConfiguration().posFileDetails;
+        static TestConfiguration EssConfig { get; set; }
```
Comment (ajasonca) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FssBatchHelper.cs:12]: EssConfig is not used, please remove.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs [hunk 3]
```diff
@@ -211,5 +176,104 @@ private void CreateIsoAndSha1ForExchangeSet(List<FssBatchFile> fileDetails, stri
...
+        {
+            ExchangeSetResponseModel exchangeSetResponseModel = await _essService.GetProductDataSinceDateTime(sinceDateTime);
+            if (!string.IsNullOrEmpty(exchangeSetResponseModel.Links.ExchangeSetBatchDetailsUri.Href))
+            {
+                string essBatchId = CommonHelper.ExtractBatchId(exchangeSetResponseModel.Links.ExchangeSetBatchDetailsUri.Href);
+                _logger.LogInformation(EventIds.BatchCreatedInESS.ToEventId(), "Batch is created by ESS successfully with BatchID - {BatchID} | {DateTime} | _X-Correlation-ID : {CorrelationId}", essBatchId, DateTime.Now.ToUniversalTime(), CommonHelper.CorrelationID);
+
+                return essBatchId;
+            }
+            else
```
Comment (ajasonca) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs:190]: Don't need else here as the if contains a return.

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs [hunk 4]
```diff
@@ -211,5 +176,104 @@ private void CreateIsoAndSha1ForExchangeSet(List<FssBatchFile> fileDetails, stri
...
+        private async Task<bool> CreatePosBatch(string downloadPath, string fileExtension, string mediaType, Batch batchType)
+        {
+            bool isBatchCreated = false;
+
+            string batchId = await _fssService.CreateBatch(mediaType, batchType);
+            IEnumerable<string> filePaths = _fileSystemHelper.GetFiles(downloadPath, fileExtension, SearchOption.TopDirectoryOnly);
+            UploadBatchFiles(filePaths, batchId);
+            bool isCommitted = await _fssService.CommitBatch(batchId, filePaths);
+
+            if (isCommitted)
```
Comment (ajasonca) [inline UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs:255]: Why not just return isCommitted and remove isBatchCreated completely?

**Review by ajasonca — CHANGES_REQUESTED** [review]
A few changes to resolve.

File: UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs [hunk 5]
```diff
@@ -11,38 +11,9 @@ public BatchResponse CreateBatch(IEnumerable<KeyValuePair<string, string>> attri
...
-                               attributes.Any(a => a.Key.ToLower() == "content" && a.Value.ToLower() == "catalogue"):
-                    batchId = EnumHelper.GetEnumDescription(BatchId.PosFmCatalogueFileBatch);
-                    break;
-                default:
-                    batchId = EnumHelper.GetEnumDescription(BatchId.PosUpdateZipBatch);
-                    break;
-            }
+            string attributeValue = attributes.Where(a => a.Key.ToLower() == "batch type").FirstOrDefault().Value.ToLower();
+            Enum.TryParse(attributeValue, true, out Batch batchType);
+            batchId = EnumHelper.GetEnumDescription(batchType);
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs]: Resolved

File: UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs [hunk 6]
```diff
@@ -67,7 +38,7 @@ public BatchDetail GetBatchDetails(string batchId)
 
             List<Models.Response.Attribute> attributes = new();
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs]: Resolved

File: UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs [hunk 7]
```diff
@@ -11,38 +11,9 @@ public BatchResponse CreateBatch(IEnumerable<KeyValuePair<string, string>> attri
...
-                    break;
-                case true when attributes.Any(a => a.Key.ToLower() == "catalogue type" && a.Value.ToLower() == "xml") &&
-                               attributes.Any(a => a.Key.ToLower() == "content" && a.Value.ToLower() == "catalogue"):
-                    batchId = EnumHelper.GetEnumDescription(BatchId.PosFmCatalogueFileBatch);
-                    break;
-                default:
-                    batchId = EnumHelper.GetEnumDescription(BatchId.PosUpdateZipBatch);
-                    break;
-            }
+            string attributeValue = attributes.Where(a => a.Key.ToLower() == "batch type").FirstOrDefault().Value.ToLower();
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.FmEssFssMock.API/UKHO.FmEssFssMock.API/Services/FileShareService.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/FunctionalTests/POSEndToEndFunctionalScenarios.cs [hunk 8]
```diff
@@ -38,32 +41,62 @@ public async Task WhenExtractedZipAndGeneratedISOAndSha1Files_ThenBatchesAreCrea
...
 
         [Test]
-        public async Task WhenICallFileDownloadEndpointWithValidBatchId_ThenALargeMediaIsoAndSha1FilesAreGenerated()
+        public async Task WhenICallFileDownloadEndpointForFullAVCSExchangeSetWithBatchId_ThenIsoAndSha1FilesAreDownloaded()
         {
-            DownloadedFolderPath = await FileContentHelper.CreateExchangeSetFileForIsoAndSha1Files(posDetails.IsoSha1BatchId, FssJwtToken);
+            HttpResponseMessage apiResponse = await GetBatchDetails.GetBatchDetailsEndpoint(FSSAuth.BaseUrl, posDetails.IsoSha1BatchId);
+            Assert.That((int)apiResponse.StatusCode, Is.EqualTo(200), $"Incorrect status code is returned {apiResponse.StatusCode}, instead of the expected status 200.");
+
+            var batchDetailsResponse = await apiResponse.DeserializeAsyncResponse();
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/FunctionalTests/POSEndToEndFunctionalScenarios.cs]: Yes, its response is used in line# 65

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FileContentHelper.cs [hunk 9]
```diff
@@ -14,43 +17,55 @@ public static async Task<List<string>> CreateExchangeSetFileForLargeMedia(string
...
-            File.Delete(Path.Combine(path, fullFileName + ".iso.sha1"));
-            File.Delete(Path.Combine(path, fullFileName));
         }
 
-        public static async Task<List<string>> DownloadAndExtractExchangeSetZipFileForLargeMedia(string batchId, string fssJwtToken)
+        public static async Task<List<string>> DownloadAndExtractExchangeSetZipFileForLargeMedia(string batchId, string fssJwtToken, dynamic batchDetailsResponse)
         {
-            List<string> downloadFolderPath = new();
+            List<string> downloadFolderPath = new List<string>();
+            string zipFileName;
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FileContentHelper.cs]: Resolved

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FssBatchHelper.cs [hunk 10]
```diff
@@ -7,39 +7,49 @@ namespace UKHO.PeriodicOutputService.API.FunctionalTests.Helpers
     public static class FssBatchHelper
     {
         private static FssApiClient FssApiClient { get; set; }
+        static FSSApiConfiguration FSSAuth = new TestConfiguration().FssConfig;
```
[the remark this answers is not in the document]
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Removed

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FssBatchHelper.cs (hunk 2 above)
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.API.FunctionalTests/Helpers/FssBatchHelper.cs]: Removed

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs (hunk 3 above)
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs]: Removed

File: UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs (hunk 4 above)
Comment (sumod13943) [reply UKHO.PeriodicOutputService/UKHO.PeriodicOutputService.Fulfilment/Services/FulfilmentDataService.cs]: Resolved

**Review by ajasonca — APPROVED** [review]
Looks good thanks.

**Pull request merged**